File: XMUOJ/XMU006/main.py
import sys

cur = input()
n = len(cur)
cur = int(cur, 2)
tar = int(input(), 2)

min_times = sys.maxsize

for p in range(2):
    times = 0
    next = p
    lock = cur

    for i in range(n):
        if next:
            times += 1
            if i > 0:
                lock ^= (1 << (i - 1))
            if i < n - 1:
                lock ^= (1 << (i + 1))
            lock ^= (1 << i)
        next = (lock >> i) & 1 != (tar >> i) & 1

    if lock == tar:
        min_times = min(min_times, times)

if min_times == sys.maxsize:
    print("impossible")
else:
    print(min_times)

File: XMUOJ/XMU006/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Ref: https://www.bilibili.com/video/BV1Zi4y1N7dU/

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int minTimes = 114514;
    string tmp;
    cin >> tmp;
    bitset<32> cur(tmp);
    cin >> tmp;
    bitset<32> tar(tmp);

    int n = tmp.size();

    for (int p = 0; p < 2; p++)
    {
        int times = 0, next = p;
        bitset<32> lock = cur;

        for (int i = 0; i < n; i++)
        {
            if (next)
            {
                times++;
                if (i > 0)  lock.flip(i - 1);
                if (i < n - 1)  lock.flip(i + 1);
                lock.flip(i);
            }
            next = (lock[i] != tar[i]);
        }
        if (lock == tar)
            minTimes = min(minTimes, times);
    }


    if (minTimes == 114514)
        cout << "impossible";
    else
        cout << minTimes;

    return 0;
}
